typing/typecore.ml

```diff
@@ -4417,15 +4358,19 @@ and type_label_exp create env loc ty_expected
     with exn when maybe_expansive arg -> try
       (* Try to retype without propagating ty_arg, cf PR#4862 *)
       Option.iter Btype.backtrack snap;
-      begin_def ();
-      let arg = type_exp env sarg in
-      end_def ();
-      lower_contravariant env arg.exp_type;
-      begin_def ();
-      let arg = {arg with exp_type = instance arg.exp_type} in
```

Comment:
Was this `instance` call useful/necessary? We just type-checked `arg` at the same level (since the call to `type_exp` we called `end_def` followed by `begin_def` again), and we did not generalize in between, only lowered contravariant variables.

Reply:
This indeed does not seem to be necessary.
This may be a leftover from a previous approach to `lower_contravariant` that simultaneously generalized the structure of the type.
An alternative (tested) version of the code fixing this would be:
```ocaml
      let arg, _ =
        with_local_level (fun () -> type_exp env sarg, instance ty_arg)
        ~post: begin fun (arg, ty_arg) ->
          lower_contravariant env arg.exp_type;
          unify_exp env arg ty_arg;
          generalize_and_check_univars env "field value" arg label.lbl_arg vars
        end;
      in
      {arg with exp_type = instance arg.exp_type}
```
This is orthogonal to the current PR, but do you think it should be included?

Reply:
I prefer the code you show above to the one in the PR, and I agree that it should be equivalent. If you want to include it in the PR, please feel free; if you want to submit it as a later PR, that is also okay with me. But I think that we want the change in trunk in the medium term.